ci/builds/checkers.sh

```diff
@@ -0,0 +1,167 @@
+  rm -f "${tmp}"
+}
+export -f sed_edit
+
+# The list of files to check.
+#
+# By default, we check all files in the repository tracked by `git`. To check
+# only the files that have changed in a development branch, set
+# `GOOGLE_CLOUD_CPP_FAST_CHECKERS=1`.
+git_files() {
+  if [ -z "${GOOGLE_CLOUD_CPP_FAST_CHECKERS-}" ]; then
+    git ls-files "${@}"
+  else
+    git diff main --name-only --diff-filter=d "${@}"
+  fi
+}
+
+# This controls the output format from bash's `time` command, which we use
+# below to time blocks of the script. A newline is automatically included.
+readonly TIMEFORMAT="... %R seconds"
+
+# Use the printf command rather than the shell builtin, which avoids issues
+# with bash sometimes buffering output from its builtins. See
+# https://github.com/googleapis/google-cloud-cpp/issues/4152
+enable -n printf
+
+# Applies whitespace fixes in text files, unless they request no edits. The
+# `[D]` character class makes this file not contain the target text itself.
+printf "%-50s" "Running whitespace fixes:" >&2
+time {
+  # Removes trailing whitespace on lines
+  expressions=("-e" "'s/[[:blank:]]\+$//'")
+  # Removes trailing blank lines (see http://sed.sourceforge.net/sed1line.txt)
+  expressions+=("-e" "':x;/^\n*$/{\$d;N;bx;}'")
+  # Adds a trailing newline if one doesn't already exist
+  expressions+=("-e" "'\$a\'")
+  git_files -z | grep -zv 'googleapis.patch$' |
+    grep -zv '\.gz$' |
+    grep -zv '\.pb$' |
+    grep -zv '\.png$' |
+    grep -zv '\.raw$' |
+    grep -zv '\.flac$' |
+    (xargs -r -0 grep -ZPL "\b[D]O NOT EDIT\b" || true) |
+    xargs -r -P "$(nproc)" -n 50 -0 bash -c "sed_edit ${expressions[*]} \"\$0\" \"\$@\""
+}
+
+# Apply shfmt to format all shell scripts
+printf "%-50s" "Running shfmt:" >&2
+time {
+  git_files -z -- '*.sh' | xargs -r -P "$(nproc)" -n 50 -0 shfmt -w
+}
+
+# Apply buildifier to fix the BUILD and .bzl formatting rules.
+#    https://github.com/bazelbuild/buildtools/tree/master/buildifier
+printf "%-50s" "Running buildifier:" >&2
+time {
+  git_files -z -- '*.BUILD' '*.bzl' '*.bazel' |
+    xargs -r -P "$(nproc)" -n 50 -0 buildifier -mode=fix
+}
+
+# The version of clang-format is important, different versions have slightly
+# different formatting output (sigh).
+printf "%-50s" "Running clang-format:" >&2
+time {
+  git_files -z -- '*.h' '*.cc' '*.proto' |
+    xargs -r -P "$(nproc)" -n 50 -0 clang-format -i
+}
+
+# Creating a virtual environment and installing the correct programs locally
+printf "%-50s" "Installing Python packages:" >&2
+VENV_NAME="env"
+# List of packages to install
+PACKAGES=(
+  mdformat
+  cmake-format
+)
+time {
+  # Check if the virtual environment already exists.
+  if [[ ! -d "$VENV_NAME" ]]; then
+    python3 -m venv "$VENV_NAME"
+  fi
+  source "$VENV_NAME/bin/activate"
+
+  # Install packages, skipping existing ones
+  for package in "${PACKAGES[@]}"; do
+    if [[ ! $(pip list | grep "^$package") ]]; then
+      pip install "$package"
+    fi
+  done
```

Comment:
I think this should work too:

```suggestion
  pip install -q "${PACKAGES[@]}"
```